### PR TITLE
[WIP] Namecheap provider added

### PR DESCRIPTION
**Pull request by HX-Rd**
Closed my first pull request, was missing the added vendor dependency.
Any comments on the code or the implementation appreciated.

### CODE REVIEW

File: builtin/providers/namecheap/provider.go [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+	return &schema.Provider{
+		Schema: map[string]*schema.Schema{
+			"username": &schema.Schema{
+				Type:        schema.TypeString,
+				Required:    true,
+				DefaultFunc: schema.EnvDefaultFunc("NAMECHEAP_USERNAME", nil),
+				Description: "A registered username for namecheap",
+			},
+
+			"apiuser": &schema.Schema{
```
Comment (jen20) [inline builtin/providers/namecheap/provider.go]: This might be better as `api_user`?

File: builtin/providers/namecheap/provider.go [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+			},
+
+			"ip": &schema.Schema{
+				Type:        schema.TypeString,
+				Required:    true,
+				DefaultFunc: schema.EnvDefaultFunc("NAMECHEAP_IP", nil),
+				Description: "IP addess of the machine running terraform",
+			},
+
+			"usesandbox": &schema.Schema{
```
Comment (jen20) [inline builtin/providers/namecheap/provider.go]: This one might be better as `use_sandbox`?

**Comment by jen20** [comment]
Hi @HX-Rd! I left a couple of questions inline on the diff, but right now the thing preventing a merge here is the lack of license on the dependency which is vendored. Terraform is MPLv2 licensed, so we need our dependencies to be compatible with that. It looks like the dependency is your own work (based on the namespace) - can you consider adding a license and updating this pull request?

In addition, can you consider squashing this into one commit and rebasing to master? With the comments in line addressed, this then LGTM! Thanks!

File: builtin/providers/namecheap/resource_namecheap_record.go [hunk 3]
```diff
@@ -0,0 +1,149 @@
...
+import (
+	"fmt"
+	"strconv"
+	"sync"
+
+	"github.com/HX-Rd/namecheap"
+	"github.com/hashicorp/terraform/helper/schema"
+)
+
+var mutex = &sync.Mutex{}
```
Comment (jen20) [inline builtin/providers/namecheap/resource_namecheap_record.go]: Is there something about the remote API that makes it necessary to take a mutex here?

**Comment by HX-Rd** [comment]
Hi @jen20 
changing usesandbox and apiuser to use_sandbox and api_user sounds good to me.
I was also thinking about changing couple of variable name on the resource it self.
1. hostname to name ( like in the dme_record resource )
2. recordType to type
3. if snake case is the way to go then maybe mxPref should be mx_pref

Then i was also thinking about adding a output parameter "hostname" (the FQDN of the record like in the dnsimple_record ).

About the mutex. I'm afraid that is it necessary be course of the namecheap api. The api does not support updating individual records so i achieve this by first getting all the records with a call to GetHost method and then add the record to that list and update it with SetHost. So there is a race condition if we are working with multiple records at the same time. Maybe is should add a comment to explain this in the code. Anyways i will try to send in a new pull request soon with these changes and i will add the license and squash it to one commit. Thanks for the comments!

**Pull request closed without merging**